classes/class-rest-primary-domain-controller.php

```diff
@@ -0,0 +1,233 @@
+		$this->rest_base         = 'primary';
+	}
+
+	/**
+	 * Register the routes for the objects of the controller.
+	 */
+	public function register_routes() {
+		register_rest_route( $this->parent_namespace, $this->rest_base, array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_item' ),
+				'permission_callback' => array( $this, 'get_item_permissions_check' ),
+			),
+			array(
+				'methods'             => WP_REST_Server::CREATABLE,
+				'callback'            => array( $this, 'create_item' ),
+				'permission_callback' => array( $this, 'create_item_permissions_check' ),
+				'args'                => array(
+					'mapping' => array(
+						'sanitize_callback' => 'absint',
+						'required'          => true,
+					),
+				),
+			),
+
+			'schema' => array( $this, 'get_public_item_schema' ),
+		) );
+	}
+
+	/**
+	 * Check if a given request has access to get a specific item.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function get_item_permissions_check( $request ) {
+		return $this->parent_controller->get_item_permissions_check( $request );
+	}
+
+	/**
+	 * Get one item from the collection.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_item( $request ) {
+		return $this->prepare_item_for_response( null, $request );
+	}
+
+	/**
+	 * Check if a given request has access to create items.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_Error|boolean
+	 */
+	public function create_item_permissions_check( $request ) {
+		return $this->get_item_permissions_check( $request );
+	}
+
+	/**
+	 * Create one item from the collection.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function create_item( $request ) {
+		$mapping_id = $request->get_param( 'mapping' );
+		$mapping    = Mapping::get( $mapping_id );
+
+		if ( is_wp_error( $mapping ) ) {
+			return $mapping;
+		}
+
+		if ( null === $mapping ) {
+			return new WP_Error( __( 'You must supply a valid mapping ID to set as the primary domain via the `mapping` parameter.', 'mercator' ) );
+		}
+
+		$result = $mapping->make_primary();
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		return $this->prepare_item_for_response( null, $request );
+	}
+
+
+	/**
+	 * Prepare the item for the REST response.
+	 *
+	 * @param mixed           $item    WordPress representation of the item.
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response $response
+	 */
+	public function prepare_item_for_response( $item, $request ) {
+
+		$blog = $this->parent_controller->get_blog_id( $request );
+		$item = get_site( $blog );
+
+		if ( is_wp_error( $item ) || is_null( $item ) ) {
+			return new WP_REST_Response( $item );
```

Comment:
`rest_ensure_response()` may be better here, which will pass through the error correctly.